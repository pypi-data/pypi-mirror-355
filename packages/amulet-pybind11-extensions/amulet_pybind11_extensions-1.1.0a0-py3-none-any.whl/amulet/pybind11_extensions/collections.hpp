#pragma once
#include <iterator>
#include <ranges>

#include <pybind11/pybind11.h>

// I have found cases where I want to accept or return a python object matching a collections.abc class.
// This extension adds subclasses of py::object with type hints for the collections.abc classes.
// This allows C++ functions to accept or return python objects that match the collection.abc classes.
// Note that these are handled in the same way as py::object thus there is no type validation.

namespace Amulet {
namespace pybind11_extensions {
    namespace collections {
        template <typename T>
        class Iterator : public pybind11::object {
        public:
            PYBIND11_OBJECT_DEFAULT(Iterator, object, PyIter_Check)

            using iterator_category = std::input_iterator_tag;
            using difference_type = pybind11::ssize_t;
            using value_type = T;

            Iterator& operator++()
            {
                advance();
                return *this;
            }

            Iterator operator++(int)
            {
                auto rv = *this;
                advance();
                return rv;
            }

            T operator*() const
            {
                return obj().template cast<T>();
            }

            std::unique_ptr<T> operator->() const
            {
                return std::make_unique<T>(obj().template cast<T>());
            }

            static Iterator sentinel() { return {}; }

            friend bool operator==(const Iterator& a, const Iterator& b) { return a.obj().ptr() == b.obj().ptr(); }
            friend bool operator!=(const Iterator& a, const Iterator& b) { return a.obj().ptr() != b.obj().ptr(); }

        private:
            void advance()
            {
                value = pybind11::reinterpret_steal<pybind11::object>(PyIter_Next(m_ptr));
                if (value.ptr() == nullptr && PyErr_Occurred()) {
                    throw pybind11::error_already_set();
                }
            }

            const pybind11::object& obj() const
            {
                if (m_ptr && !value.ptr()) {
                    auto& self = const_cast<Iterator&>(*this);
                    self.advance();
                }
                return value;
            }

        private:
            pybind11::object value = {};
        };
        static_assert(std::input_iterator<Iterator<int>>);

        template <typename T>
        class Iterable : public pybind11::object {
            PYBIND11_OBJECT_DEFAULT(Iterable, object, pybind11::detail::PyIterable_Check)

            Iterator<T> begin() const
            {
                return Iterator<T>(pybind11::object::begin());
            }
            Iterator<T> end() const
            {
                return Iterator<T>(pybind11::object::end());
            }
        };
        static_assert(std::ranges::input_range<Iterable<int>>);
        static_assert(std::convertible_to<std::ranges::range_value_t<Iterable<int>>, const int&>);

        template <typename T>
        class Sequence : public pybind11::object {
            PYBIND11_OBJECT_DEFAULT(Sequence, object, PySequence_Check)

            Iterator<T> begin() const
            {
                return Iterator<T>(pybind11::object::begin());
            }

            Iterator<T> end() const
            {
                return Iterator<T>(pybind11::object::end());
            }

            Py_ssize_t size() const
            {
                return PyObject_Size(ptr());
            }
        };

        template <typename K, typename V>
        class Mapping : public pybind11::object {
            PYBIND11_OBJECT_DEFAULT(Mapping, object, PyMapping_Check)
        };

        template <typename K, typename V>
        class MutableMapping : public pybind11::object {
            PYBIND11_OBJECT_DEFAULT(MutableMapping, object, PyMapping_Check)
        };

        template <typename K>
        class KeysView : public pybind11::object {
            PYBIND11_OBJECT_DEFAULT(KeysView, object, PyObject_Type)
        };

        template <typename V>
        class ValuesView : public pybind11::object {
            PYBIND11_OBJECT_DEFAULT(ValuesView, object, PyObject_Type)
        };

        template <typename K, typename V>
        class ItemsView : public pybind11::object {
            PYBIND11_OBJECT_DEFAULT(ItemsView, object, PyObject_Type)
        };
    } // namespace collections
} // namespace pybind11_extensions
} // namespace Amulet

namespace pybind11 {
namespace detail {
    template <typename T>
    struct handle_type_name<Amulet::pybind11_extensions::collections::Iterator<T>> {
        static constexpr auto name = const_name("collections.abc.Iterator[") + make_caster<T>::name + const_name("]");
    };

    template <typename T>
    struct handle_type_name<Amulet::pybind11_extensions::collections::Iterable<T>> {
        static constexpr auto name = const_name("collections.abc.Iterable[") + make_caster<T>::name + const_name("]");
    };

    template <typename T>
    struct handle_type_name<Amulet::pybind11_extensions::collections::Sequence<T>> {
        static constexpr auto name = const_name("collections.abc.Sequence[") + make_caster<T>::name + const_name("]");
    };

    template <typename K, typename V>
    struct handle_type_name<Amulet::pybind11_extensions::collections::Mapping<K, V>> {
        static constexpr auto name = const_name("collections.abc.Mapping[") + make_caster<K>::name + const_name(", ") + make_caster<V>::name + const_name("]");
    };

    template <typename K, typename V>
    struct handle_type_name<Amulet::pybind11_extensions::collections::MutableMapping<K, V>> {
        static constexpr auto name = const_name("collections.abc.MutableMapping[") + make_caster<K>::name + const_name(", ")
            + make_caster<V>::name + const_name("]");
    };

    template <typename K>
    struct handle_type_name<Amulet::pybind11_extensions::collections::KeysView<K>> {
        static constexpr auto name = const_name("collections.abc.KeysView[") + make_caster<K>::name + const_name("]");
    };

    template <typename V>
    struct handle_type_name<Amulet::pybind11_extensions::collections::ValuesView<V>> {
        static constexpr auto name = const_name("collections.abc.ValuesView[") + make_caster<V>::name + const_name("]");
    };

    template <typename K, typename V>
    struct handle_type_name<Amulet::pybind11_extensions::collections::ItemsView<K, V>> {
        static constexpr auto name = const_name("collections.abc.ItemsView[") + make_caster<K>::name + const_name(", ")
            + make_caster<V>::name + const_name("]");
    };
}
}
