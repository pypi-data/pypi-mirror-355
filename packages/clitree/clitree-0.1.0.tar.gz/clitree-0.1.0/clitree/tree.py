from typing import Any, Callable, Iterable, TypeVar

Node = Any
T = TypeVar("T")
NodeValueGetter = Callable[[Node], T]


def _get_value(node: Node, key: str | NodeValueGetter[T]) -> T | None:
    """
    Get the value of a key or attribute from a node.

    Args:
        node: The node to get the value from
        key: The key or attribute to get the value from. Can be a string key or a callable

    Returns:
        The value of the key or attribute
    """
    if callable(key):
        return key(node)
    if hasattr(node, "get"):
        return node.get(key)
    return getattr(node, key, None)


def _get_name(node: Node, name: str | NodeValueGetter[str]) -> str:
    result = _get_value(node, name)
    if result is None:
        raise ValueError(f"Node name resolved to None: {node}")
    return str(result)


def _get_children(
    node: Node, children: str | NodeValueGetter[Iterable[Node] | None]
) -> Iterable[Node] | None:
    result = _get_value(node, children)
    if result is None:
        return None
    return result


def _build_tree(
    node: Node,
    name: str | NodeValueGetter[str],
    children: str | NodeValueGetter[Iterable[Node] | None],
    prefix: str = "",
) -> str:
    node_name = _get_name(node, name)
    node_children = _get_children(node, children) or []
    children_list = list(node_children)

    result = [node_name]

    for i, child in enumerate(children_list):
        is_last_child = i == len(children_list) - 1
        connector = "└── " if is_last_child else "├── "
        child_prefix = prefix + ("    " if is_last_child else "│   ")
        result.append(
            prefix + connector + _build_tree(child, name, children, child_prefix)
        )

    return "\n".join(result)


def tree(
    data: Node,
    name: str | NodeValueGetter[str] = "name",
    children: str | NodeValueGetter[Iterable[Node] | None] = "children",
) -> str:
    """
    Generate a tree-like string representation of a hierarchical data structure.

    Args:
        data: The root node of the tree structure
        name: How to extract the name of a node. Can be a string key or a callable
        children: How to extract children of a node. Can be a string key or a callable

    Returns:
        A string representation of the tree structure

    Raises:
        ValueError: If the name resolves to None
        TypeError: If children are not iterable
    """
    return _build_tree(data, name, children)
