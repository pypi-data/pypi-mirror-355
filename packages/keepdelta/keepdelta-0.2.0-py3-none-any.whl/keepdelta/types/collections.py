from copy import deepcopy

from keepdelta.check import CheckConflict
from keepdelta.config import keys
from keepdelta.types.primitives import (
    DeltaBool,
    DeltaComplex,
    DeltaFloat,
    DeltaInt,
    DeltaStr,
)


class Delta:
    """
    Handle deltas for all variables
    """

    @staticmethod
    def create(old, new):
        """
        Create delta for the variables
        """
        type_new = type(new)  # Avoid multiple calls to type()
        type_old = type(old)  # Avoid multiple calls to type()
        if type_old == type_new:  # Same type
            if old != new:  # Unequal and same type
                if type_new is bool:  # bool
                    delta = DeltaBool.create(old, new)
                elif type_new is complex:  # complex
                    delta = DeltaComplex.create(old, new)
                elif type_new is float:  # float
                    delta = DeltaFloat.create(old, new)
                elif type_new is int:  # int
                    delta = DeltaInt.create(old, new)
                elif type_new is str:  # str
                    CheckConflict.check_str(new)
                    delta = DeltaStr.create(old, new)
                elif type_new is dict:  # dict
                    CheckConflict.check_dict(new)
                    delta = DeltaDict.create(old, new)
                elif type_new is list:  # list
                    CheckConflict.check_list(new)
                    delta = DeltaList.create(old, new)
                elif type_new is set:  # set
                    CheckConflict.check_set(new)
                    delta = DeltaSet.create(old, new)
                elif type_new is tuple:  # tuple
                    CheckConflict.check_tuple(new)
                    delta = DeltaTuple.create(old, new)
                else:
                    raise ValueError("Variable type not supported: ", type(old))
            else:  # Equal and same type
                delta = keys["nothing"]
        else:  # Not same type
            # Check `old`
            if type_old is str:  # str
                CheckConflict.check_str(old)
            elif type_old is dict:  # dict
                CheckConflict.check_dict(old)
            elif type_old is list:  # list
                CheckConflict.check_list(old)
            elif type_old is set:  # set
                CheckConflict.check_set(old)
            elif type_old is tuple:  # tuple
                CheckConflict.check_tuple(old)
            # Check `new`
            if type_new is str:  # str
                CheckConflict.check_str(new)
            elif type_new is dict:  # dict
                CheckConflict.check_dict(new)
            elif type_new is list:  # list
                CheckConflict.check_list(new)
            elif type_new is set:  # set
                CheckConflict.check_set(new)
            elif type_new is tuple:  # tuple
                CheckConflict.check_tuple(new)
            delta = new  # Rewrite type
        return delta

    @staticmethod
    def apply(old, delta):
        """
        Apply delta to the variable
        """
        if delta == keys["nothing"]:  # Equal and same type
            new = deepcopy(old)
        else:
            if type(old) is bool and type(delta) is bool:  # bool
                new = DeltaBool.apply(old, delta)
            elif type(old) is complex and type(delta) is complex:  # complex
                new = DeltaComplex.apply(old, delta)
            elif type(old) is float and type(delta) is float:  # float
                new = DeltaFloat.apply(old, delta)
            elif type(old) is int and type(delta) is int:  # int
                new = DeltaInt.apply(old, delta)
            elif type(old) is str and type(delta) is str:  # str
                new = DeltaStr.apply(old, delta)
            elif type(old) is dict and type(delta) is dict:  # dict
                new = DeltaDict.apply(old, delta)
            elif type(old) is list and type(delta) is dict:  # list
                new = DeltaList.apply(old, delta)
            elif type(old) is set and type(delta) is dict:  # set
                new = DeltaSet.apply(old, delta)
            elif type(old) is tuple and type(delta) is dict:  # tuple
                new = DeltaTuple.apply(old, delta)
            else:  # Variable type not recognized or None
                new = deepcopy(delta)
        return new


class DeltaDict:
    """
    Handle deltas for dict variables
    """

    @staticmethod
    def create(old: dict, new: dict) -> dict:
        """
        Create delta for dict variables
        """
        delta = {}
        old_keys = old.keys()
        new_keys = new.keys()
        keys_in_both = list(set(old_keys) & set(new_keys))
        keys_in_old_not_in_new = list(set(old_keys) - set(new_keys))
        keys_in_new_not_in_old = list(set(new_keys) - set(old_keys))
        for key in keys_in_both:
            delta_value = Delta.create(old[key], new[key])
            if delta_value != keys["nothing"]:
                delta[key] = delta_value
        for key in keys_in_old_not_in_new:
            delta[key] = keys["delete"]
        for key in keys_in_new_not_in_old:
            delta_value = Delta.create(None, new[key])
            if delta_value != keys["nothing"]:
                delta[key] = delta_value
        return delta

    @staticmethod
    def apply(old: dict, delta: dict) -> dict:
        """
        Apply delta to the dict variable
        """
        new = deepcopy(old)
        for key in delta:
            delta_value = delta[key]
            if key in old:
                if delta_value == keys["delete"]:
                    del new[key]
                else:
                    new[key] = Delta.apply(old[key], delta_value)
            else:
                new[key] = delta_value
        return new


class DeltaList:
    """
    Handle deltas for list variables
    """

    @staticmethod
    def create(old: list, new: list) -> dict:
        """
        Create delta for list variables
        """
        old = DeltaList._list_to_dict(old)
        new = DeltaList._list_to_dict(new)
        return DeltaDict.create(old, new)

    @staticmethod
    def apply(old: list, delta: dict) -> list:
        """
        Apply delta to the list variable
        """
        old_dict = DeltaList._list_to_dict(old)
        new_dict = DeltaDict.apply(old_dict, delta)
        return DeltaList._dict_to_list(new_dict)

    @staticmethod
    def _list_to_dict(input: list) -> dict:
        """
        Convert list to dictionary
        """
        result = {}
        for i, item in enumerate(input):
            result[i] = item
        return result

    @staticmethod
    def _dict_to_list(input: dict) -> list:
        """
        Convert dictionary to list
        """
        result = None
        if input is not None:
            result = []
            for i in input:
                if input[i] is not None:
                    result.append(input[i])
        return result


class DeltaTuple:
    """
    Handle deltas for tuple variables
    """

    @staticmethod
    def create(old: tuple, new: tuple) -> dict:
        """
        Create delta for tuple variables
        """
        old_list = list(old)
        new_list = list(new)
        return DeltaList.create(old_list, new_list)

    @staticmethod
    def apply(old: tuple, delta: dict) -> tuple:
        """
        Apply delta to the tuple variable
        """
        old_list = list(old)
        new_list = DeltaList.apply(old_list, delta)
        return tuple(new_list)


class DeltaSet:
    """
    Handle deltas for set variables
    """

    @staticmethod
    def create(old: set, new: set) -> dict:
        """
        Create delta for set variables
        """
        delta = {}
        # Add annotation
        annotated_old = DeltaSet.add_annotation(old)
        annotated_new = DeltaSet.add_annotation(new)
        # Elements to add
        elements_to_add = annotated_new - annotated_old
        if len(elements_to_add) > 0:
            elements_to_add = DeltaSet.remove_annotation(elements_to_add)
            delta[keys["add to set"]] = elements_to_add
        # Elements to remove
        elements_to_remove = annotated_old - annotated_new
        if len(elements_to_remove) > 0:
            elements_to_remove = DeltaSet.remove_annotation(elements_to_remove)
            delta[keys["remove from set"]] = elements_to_remove
        return delta

    @staticmethod
    def apply(old: set, delta: dict) -> set:
        """
        Apply delta to the set variable
        """
        # Distinguish between bool and other types
        old_bools = {x for x in old if type(x) is bool}
        add_bools = {x for x in delta.get(keys["add to set"], set()) if type(x) is bool}
        remove_bools = {
            x for x in delta.get(keys["remove from set"], set()) if type(x) is bool
        }
        old_non_bools = old - old_bools
        add_non_bools = delta.get(keys["add to set"], set()) - add_bools
        remove_non_bools = delta.get(keys["remove from set"], set()) - remove_bools
        updated_non_bools = (old_non_bools | add_non_bools) - remove_non_bools
        updated_bools = (old_bools | add_bools) - remove_bools
        return updated_non_bools | updated_bools

    @staticmethod
    def add_annotation(var: set) -> set:
        """
        Add type of each variable
        """
        return {(x, str(type(x))) for x in var}

    @staticmethod
    def remove_annotation(var: set) -> set:
        """
        Remove type of each variable
        """
        return {x[0] for x in var}


if __name__ == "__main__":
    old_var = {"a": 1, "b": 2, "c": {"d": 5, "e": 2}}
    new_var = {"a": 3, "c": {"d": 5, "e": 3}}
    expected_delta = {"a": 2, "b": keys["delete"], "c": {"e": 1}}

    # Create delta
    delta = Delta.create(old_var, new_var)
    print("Delta:", delta)
    print("Test delta creation: ", delta == expected_delta)

    # Apply delta
    var = Delta.apply(old_var, delta)
    print("Reconstructed variable:", var)
    print("Test delta application: ", var == new_var)
