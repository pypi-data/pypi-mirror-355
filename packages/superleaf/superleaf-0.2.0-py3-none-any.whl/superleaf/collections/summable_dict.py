from typing import Self


class SummableDict(dict):
    """A dictionary-like object that supports addition and subtraction of values."""

    def __add__(self, other) -> Self:
        summed = self.copy()
        if isinstance(other, dict):
            for k, v in other.items():
                if k in summed:
                    summed[k] = summed[k] + v
                else:
                    summed[k] = v
        else:
            for k, v in self.items():
                summed[k] = summed[k] + other
        return summed

    def __iadd__(self, other) -> Self:
        if isinstance(other, dict):
            for k, v in other.items():
                if k in self:
                    self[k] = self[k] + v
                else:
                    self[k] = v
        else:
            for k, v in self.items():
                self[k] = self[k] + other
        return self

    def __radd__(self, other) -> Self:
        if other == 0:
            return self.copy()
        elif isinstance(other, dict):
            return SummableDict(other) + self
        else:
            summed = self.copy()
            for k, v in self.items():
                summed[k] = other + v
            return summed

    def __neg__(self) -> Self:
        return SummableDict({k: -v for k, v in self.items()})

    def __sub__(self, other) -> Self:
        if isinstance(other, dict):
            return self + -SummableDict(other)
        else:
            return self + -other

    def __isub__(self, other) -> Self:
        if isinstance(other, dict):
            for k, v in other.items():
                if k in self:
                    self[k] = self[k] - v
                else:
                    self[k] = -v
        else:
            for k, v in self.items():
                self[k] = self[k] - other
        return self

    def __rsub__(self, other):
        if other == 0:
            return self.copy()
        else:
            return other + -self

    def copy(self):
        return SummableDict(super().copy())
