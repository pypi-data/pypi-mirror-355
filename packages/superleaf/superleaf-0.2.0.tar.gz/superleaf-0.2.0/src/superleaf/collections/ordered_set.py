import itertools
from typing import Generic, Iterable, Iterator, Self, TypeVar

T = TypeVar('T')


class OrderedSet(Generic[T]):
    """Similar interface to the native set class, but with item order maintained, and expanded functionality, including
    addition and summation. Implemented by storing the set items as keys in an internal dict."""

    def __init__(self, items: Iterable[T] = None):
        self._dict: dict[T, None] = dict(zip(items, itertools.repeat(None))) if items is not None else {}

    @property
    def _items(self) -> list[T]:
        return list(self._dict.keys())

    def __iter__(self) -> Iterator[T]:
        return iter(self._items)

    def copy(self) -> Self:
        return self.__class__(self._items)

    def union(self, other: Iterable[T]) -> Self:
        return self.__class__(itertools.chain(self, other))

    def add(self, item: T) -> Self:
        self._dict[item] = None

    def intersection(self, other: Iterable[T]) -> Self:
        return self.__class__(filter(lambda x: x in other, self))

    def __add__(self, other: Iterable[T]) -> Self:
        return self.__class__(self.union(other))

    def __radd__(self, other: Iterable[T]) -> Self:
        if other == 0:
            return self
        else:
            if not isinstance(other, type(self)):
                other = type(self)(other)
            return other + self

    def __iadd__(self, other: Iterable[T]) -> Self:
        if not isinstance(other, self.__class__):
            other = self.__class__(other)
        self._dict.update(other._dict)
        return self

    def __sub__(self, other: Iterable[T]) -> Self:
        return self.__class__(filter(lambda x: x not in other, self))

    def __isub__(self, other: Iterable[T]) -> Self:
        if not isinstance(other, self.__class__):
            other = self.__class__(other)
        for item in other:
            if item in self:
                self._dict.pop(item)
        return self

    def __contains__(self, item: T) -> bool:
        return item in self._dict

    def __eq__(self, other: Self | set[T]) -> Self:
        if isinstance(other, set):
            return set(self._items) == other
        elif isinstance(other, OrderedSet):
            return self._dict == other._dict
        else:
            return False

    def __repr__(self) -> str:
        return "{" + ", ".join([item.__repr__() for item in self._items]) + "}"

    def __len__(self) -> int:
        return len(self._dict)

    def __getitem__(self, item: int) -> T:
        return self._items[item]
