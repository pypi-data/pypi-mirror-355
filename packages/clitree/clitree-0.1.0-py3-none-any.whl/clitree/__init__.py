"""A library to draw tree structures like the CLI utility 'tree' does."""

__version__ = "0.1.0"

from .tree import tree

__all__ = ["tree"]
