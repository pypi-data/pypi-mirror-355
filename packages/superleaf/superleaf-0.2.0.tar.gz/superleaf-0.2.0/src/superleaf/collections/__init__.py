from superleaf.collections.ordered_set import OrderedSet
from superleaf.collections.summable_dict import SummableDict
