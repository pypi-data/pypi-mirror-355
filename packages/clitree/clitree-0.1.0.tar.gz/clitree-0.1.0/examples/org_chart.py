from collections import defaultdict

from clitree import tree

# List of (name, position, manager) with Spanish names
ORG_TUPLES = [
    ("Juan García", "CEO", None),
    ("Alicia Fernández", "VP Ingeniería", "Juan García"),
    ("Wendy Moreno", "VP Producto", "Juan García"),
    ("Tina Ruiz", "VP Operaciones", "Juan García"),
    ("Roberto Torres", "Director Backend", "Alicia Fernández"),
    ("Iván Martínez", "Director Frontend", "Alicia Fernández"),
    ("Pedro Reyes", "Director QA", "Alicia Fernández"),
    ("Javier Díaz", "Director UX", "Wendy Moreno"),
    ("Eva Castro", "Director PM", "Wendy Moreno"),
    ("León Morales", "Director Datos", "Wendy Moreno"),
    ("Ulises Vega", "Director RRHH", "Tina Ruiz"),
    ("Beatriz Gil", "Director Finanzas", "Tina Ruiz"),
    ("Ignacio Navarro", "Director IT", "Tina Ruiz"),
    ("Patricia Ramos", "Director Legal", "Tina Ruiz"),
    ("Carlos Blanco", "Manager Backend 1", "Roberto Torres"),
    ("Francisco Molina", "Manager Backend 2", "Roberto Torres"),
    ("Joaquín Romero", "Manager Frontend 1", "Iván Martínez"),
    ("Miriam Gómez", "Manager Frontend 2", "Iván Martínez"),
    ("Quintín Alonso", "Manager QA 1", "Pedro Reyes"),
    ("Tomás Bravo", "Manager QA 2", "Pedro Reyes"),
    ("Yara Esteban", "Manager UX 1", "Javier Díaz"),
    ("Benjamín Jiménez", "Manager UX 2", "Javier Díaz"),
    ("Federico Granados", "Manager PM 1", "Eva Castro"),
    ("Irene Jurado", "Manager PM 2", "Eva Castro"),
    ("Mayra Nieto", "Manager Datos 1", "León Morales"),
    ("Pablo Quintana", "Manager Datos 2", "León Morales"),
    ("Verónica Aguirre", "Manager RRHH 1", "Ulises Vega"),
    ("Yolanda Domínguez", "Manager RRHH 2", "Ulises Vega"),
    ("César Herrera", "Manager Finanzas 1", "Beatriz Gil"),
    ("Fiona Lara", "Manager Finanzas 2", "Beatriz Gil"),
    ("Julia Ortega", "Manager IT 1", "Ignacio Navarro"),
    ("Miguel Suárez", "Manager IT 2", "Ignacio Navarro"),
    ("Quirino Barrera", "Manager Legal 1", "Patricia Ramos"),
    ("Tatiana Escobar", "Manager Legal 2", "Patricia Ramos"),
    ("David Méndez", "IC Backend 1", "Carlos Blanco"),
    ("Elena Delgado", "IC Backend 2", "Carlos Blanco"),
    ("Gabriel Tejada", "IC Backend 3", "Francisco Molina"),
    ("Héctor Andrade", "IC Backend 4", "Francisco Molina"),
    ("Karina León", "IC Frontend 1", "Joaquín Romero"),
    ("Luis Blanco", "IC Frontend 2", "Joaquín Romero"),
    ("Nora Cordero", "IC Frontend 3", "Miriam Gómez"),
    ("Olga Herrera", "IC Frontend 4", "Miriam Gómez"),
    ("Raúl Sanz", "IC QA 1", "Quintín Alonso"),
    ("Sofía Verde", "IC QA 2", "Quintín Alonso"),
    ("Úrsula Paredes", "IC QA 3", "Tomás Bravo"),
    ("Vicente Lobo", "IC QA 4", "Tomás Bravo"),
    ("Zaira Fuentes", "IC UX 1", "Yara Esteban"),
    ("Aarón Hidalgo", "IC UX 2", "Yara Esteban"),
    ("Carla Lamas", "IC UX 3", "Benjamín Jiménez"),
    ("Damián Montes", "IC UX 4", "Benjamín Jiménez"),
    ("Gisela Hoyos", "IC PM 1", "Federico Granados"),
    ("Hugo Ibarra", "IC PM 2", "Federico Granados"),
    ("Jorge Kuri", "IC PM 3", "Irene Jurado"),
    ("Katia Lobo", "IC PM 4", "Irene Jurado"),
    ("Nicolás Olmedo", "IC Datos 1", "Mayra Nieto"),
    ("Ofelia Prieto", "IC Datos 2", "Mayra Nieto"),
    ("Rita Salas", "IC Datos 3", "Pablo Quintana"),
    ("Samuel Torres", "IC Datos 4", "Pablo Quintana"),
    ("Walter Benítez", "IC RRHH 1", "Verónica Aguirre"),
    ("Ximena Cobo", "IC RRHH 2", "Verónica Aguirre"),
    ("Zulema Díaz", "IC RRHH 3", "Yolanda Domínguez"),
    ("Alonso Figueroa", "IC RRHH 4", "Yolanda Domínguez"),
    ("Diana Iglesias", "IC Finanzas 1", "César Herrera"),
    ("Esteban Jiménez", "IC Finanzas 2", "César Herrera"),
    ("Guadalupe Lira", "IC Finanzas 3", "Fiona Lara"),
    ("Helena Molina", "IC Finanzas 4", "Fiona Lara"),
    ("Kevin Ponce", "IC IT 1", "Julia Ortega"),
    ("Laura Quintana", "IC IT 2", "Julia Ortega"),
    ("Nuria Ríos", "IC IT 3", "Miguel Suárez"),
    ("Octavio Sánchez", "IC IT 4", "Miguel Suárez"),
    ("Rocío Torres", "IC Legal 1", "Quirino Barrera"),
    ("Sergio Urrutia", "IC Legal 2", "Quirino Barrera"),
    ("Ulises Vargas", "IC Legal 3", "Tatiana Escobar"),
    ("Valeria Zúñiga", "IC Legal 4", "Tatiana Escobar"),
]


def build_org_tree(tuples):
    nodes = {}
    children = defaultdict(list)
    for name, position, manager in tuples:
        nodes[name] = {"name": f"{name} ({position})"}
        if manager:
            children[manager].append(name)
    for manager, reports in children.items():
        nodes[manager]["reports"] = [nodes[r] for r in reports]
    root = next(
        n
        for n, (name, position, manager) in zip(nodes.values(), tuples)
        if manager is None
    )
    return root


if __name__ == "__main__":
    org_chart = build_org_tree(ORG_TUPLES)
    print(tree(org_chart, children="reports"))
