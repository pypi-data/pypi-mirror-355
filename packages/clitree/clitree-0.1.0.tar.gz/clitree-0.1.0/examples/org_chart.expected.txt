Juan García (CEO)
├── Alicia Fernández (VP Ingeniería)
│   ├── Roberto Torres (Director Backend)
│   │   ├── Carlos Blanco (Manager Backend 1)
│   │   │   ├── David Méndez (IC Backend 1)
│   │   │   └── Elena Delgado (IC Backend 2)
│   │   └── Francisco Molina (Manager Backend 2)
│   │       ├── Gabriel Tejada (IC Backend 3)
│   │       └── Héctor Andrade (IC Backend 4)
│   ├── Iván Martínez (Director Frontend)
│   │   ├── Joaquín Romero (Manager Frontend 1)
│   │   │   ├── Karina León (IC Frontend 1)
│   │   │   └── Luis Blanco (IC Frontend 2)
│   │   └── Miriam Gómez (Manager Frontend 2)
│   │       ├── Nora Cordero (IC Frontend 3)
│   │       └── Olga Herrera (IC Frontend 4)
│   └── Pedro Reyes (Director QA)
│       ├── Quintín Alonso (Manager QA 1)
│       │   ├── Raúl Sanz (IC QA 1)
│       │   └── Sofía Verde (IC QA 2)
│       └── Tomás Bravo (Manager QA 2)
│           ├── Úrsula Paredes (IC QA 3)
│           └── Vicente Lobo (IC QA 4)
├── Wendy Moreno (VP Producto)
│   ├── Javier Díaz (Director UX)
│   │   ├── Yara Esteban (Manager UX 1)
│   │   │   ├── Zaira Fuentes (IC UX 1)
│   │   │   └── Aarón Hidalgo (IC UX 2)
│   │   └── Benjamín Jiménez (Manager UX 2)
│   │       ├── Carla Lamas (IC UX 3)
│   │       └── Damián Montes (IC UX 4)
│   ├── Eva Castro (Director PM)
│   │   ├── Federico Granados (Manager PM 1)
│   │   │   ├── Gisela Hoyos (IC PM 1)
│   │   │   └── Hugo Ibarra (IC PM 2)
│   │   └── Irene Jurado (Manager PM 2)
│   │       ├── Jorge Kuri (IC PM 3)
│   │       └── Katia Lobo (IC PM 4)
│   └── León Morales (Director Datos)
│       ├── Mayra Nieto (Manager Datos 1)
│       │   ├── Nicolás Olmedo (IC Datos 1)
│       │   └── Ofelia Prieto (IC Datos 2)
│       └── Pablo Quintana (Manager Datos 2)
│           ├── Rita Salas (IC Datos 3)
│           └── Samuel Torres (IC Datos 4)
└── Tina Ruiz (VP Operaciones)
    ├── Ulises Vega (Director RRHH)
    │   ├── Verónica Aguirre (Manager RRHH 1)
    │   │   ├── Walter Benítez (IC RRHH 1)
    │   │   └── Ximena Cobo (IC RRHH 2)
    │   └── Yolanda Domínguez (Manager RRHH 2)
    │       ├── Zulema Díaz (IC RRHH 3)
    │       └── Alonso Figueroa (IC RRHH 4)
    ├── Beatriz Gil (Director Finanzas)
    │   ├── César Herrera (Manager Finanzas 1)
    │   │   ├── Diana Iglesias (IC Finanzas 1)
    │   │   └── Esteban Jiménez (IC Finanzas 2)
    │   └── Fiona Lara (Manager Finanzas 2)
    │       ├── Guadalupe Lira (IC Finanzas 3)
    │       └── Helena Molina (IC Finanzas 4)
    ├── Ignacio Navarro (Director IT)
    │   ├── Julia Ortega (Manager IT 1)
    │   │   ├── Kevin Ponce (IC IT 1)
    │   │   └── Laura Quintana (IC IT 2)
    │   └── Miguel Suárez (Manager IT 2)
    │       ├── Nuria Ríos (IC IT 3)
    │       └── Octavio Sánchez (IC IT 4)
    └── Patricia Ramos (Director Legal)
        ├── Quirino Barrera (Manager Legal 1)
        │   ├── Rocío Torres (IC Legal 1)
        │   └── Sergio Urrutia (IC Legal 2)
        └── Tatiana Escobar (Manager Legal 2)
            ├── Ulises Vargas (IC Legal 3)
            └── Valeria Zúñiga (IC Legal 4)
